puzzlehq/sdk
# wasm msg gen and signing

## Motivation

This creates a msg. generation function for clients and programs as well as a `sign_message` function needed for an on-chain multisig style. See comments [here](https://linear.app/puzzlehq/issue/PUZ-262/generate-correct-message-in-dapp-for-leo-token-transfer) for additional details on thought process.

Note: Also included in this pr is a merge of testnet3 from @aleohq/sdk.

Files with wasm changes:
`wasm/src/types.rs`
`wasm/src/account/field.rs`
`wasm/src/account/signature.rs` -> `sign_message`

## Test Plan
1. Clone [this](https://github.com/puzzlehq/leo-sig-verification ) repo -> needed for signature verification
2. Fill in a .env file at the root of the above repo with the private key that you will use. Any will suffice.
3. Pop over to devtools in firenze on this [branch](https://github.com/puzzlehq/firenze/pull/328) (run `rm -rf node_modules && pnpm i && pnpm devtools`
4. Click the sign tab. Input your private key (the same one that is in your `.env` file in the leo-signature-verification repo.
5. Input a message ("place_bid 10u8" or "private_transfer 10u32 aleo1xz0uyudcr24q8nmxr8nlk82penl8jtqyfyuyz6mr5udlt0g3vyfqt895ew '{ owner: aleo1hy0uyudcr24q8nmxr8nlk82penl8jtqyfyuyz6mr5udlt0g3vyfqt9l7ew, amount: 10u128 }'")
6. Copy the fields from the devtools field and place into the inputs file of the leo signature verification [repo]( https://github.com/puzzlehq/leo-sig-verification/blob/main/inputs/signature_verification.in).
7. Copy the signature from devtools field place into the inputs file of the leo signature verification [repo]( into the inputs file of the leo signature verification [repo]( https://github.com/puzzlehq/leo-sig-verification/blob/main/inputs/signature_verification.in).inputs/signature_verification.in).
8. Run `leo clean && leo build && leo run verify_field` inside the root of the leo-signature-verification [repo] (https://github.com/puzzlehq/leo-sig-verification/blob/main)
9. See the ouput `true` to console. 

<img width="1199" alt="Screenshot 2023-10-11 at 6 37 59 PM" src="https://github.com/puzzlehq/sdk/assets/28710772/b34374c7-84f4-4691-b06f-3e862168edd4">


(Write your test plan here)
<img width="1199" alt="Screenshot 2023-10-11 at 6 08 36 PM" src="https://github.com/puzzlehq/sdk/assets/28710772/d919faf0-97ce-426f-a9d4-33d1080f8fb7">

## Related PRs



[(Firenze PR)](https://github.com/puzzlehq/firenze/pull/328)

NOTE: I did not update the aleo wasm node js package [here](https://www.npmjs.com/package/@puzzlehq/aleo-wasm-nodejs)
